flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,315 @@
+
+  /** Adds a new item to the histogram
+    *
+    * @param p value to be added
+    */
+  override def add(p: Double): Unit = {
+    require(p > lower && p < upper, p + " not in (" + lower + "," + upper + ")")
+    // search for the index where the value is just higher than p
+    val search = find(p)
+    // add the new value there, shifting everything to the right
+    data.insert(search, (p, 1))
+    // If we're over capacity or any two elements are within 1e-9 of each other, merge.
+    // This will take care of the case if p was actually equal to some value in the histogram and
+    // just increment the value there
+    mergeElements()
+  }
+
+  /** Merges the histogram with h and returns a histogram with capacity B
+    *
+    * @param h histogram to be merged
+    * @param B capacity of the resultant histogram
+    * @return Merged histogram with capacity B
+    */
+  override def merge(h: OnlineHistogram, B: Int): ContinuousHistogram = {
+    h match {
+      case temp: ContinuousHistogram => {
+        val m: Int = bins
+        val n: Int = temp.bins
+        var i, j: Int = 0
+        val mergeList = new mutable.ArrayBuffer[(Double, Int)]()
+        while (i < m || j < n) {
+          if (i >= m) {
+            mergeList += ((temp.getValue(j), temp.getCounter(j)))
+            j = j + 1
+          } else if (j >= n || getValue(i) <= temp.getValue(j)) {
+            mergeList += data.apply(i)
+            i = i + 1
+          } else {
+            mergeList += ((temp.getValue(j), temp.getCounter(j)))
+            j = j + 1
+          }
+        }
+        // the size will be brought to capacity while constructing the new histogram itself
+        val finalLower = Math.min(lower, temp.lower)
+        val finalUpper = Math.max(upper, temp.upper)
+        val ret = new ContinuousHistogram(B, finalLower, finalUpper)
+        ret.loadData(mergeList.toArray)
+        ret
+      }
+      case default =>
+        throw new RuntimeException("Only a continuous histogram is allowed to be merged with a " +
+          "continuous histogram")
+
+    }
+  }
+
+  /** Returns the qth quantile of the histogram
+    *
+    * @param q Quantile value in (0,1)
+    * @return Value at quantile q
+    */
+  def quantile(q: Double): Double = {
+    require(bins > 0, "Histogram is empty")
+    require(q > 0 && q < 1, "Quantile must be between 0 and 1")
+    val wantedSum = (q * total).round.toInt
+    var currSum = count(getValue(0))
+
+    if (wantedSum < currSum) {
+      require(lower > -MaxValue, "Set a lower bound before proceeding")
+      return Math.sqrt(2 * wantedSum * Math.pow(getValue(0) - lower, 2) / getCounter(0)) + lower
+    }
+
+    /** Walk the histogram to find sums at every bin value
+      * As soon as you hit the interval where you should be
+      * Walk along the trapezoidal line
+      * This leads to solving a quadratic equation.
+      */
+    for (i <- 1 to bins - 1) {
+      val tmpSum = count(getValue(i))
```

Comment:
This is painfully slow because count has a complexity of `O(n)`, thus making the `quantile` operation's complexity effectively quadratic. Why don't you simply add up the `getCounter` values yourself and keep track of the current accumulated value?